src/Juvix/Compiler/Core/Transformation/Optimize/SpecializeArgs.hs

```diff
@@ -0,0 +1,206 @@
+                        tyargsNum = length (takeWhile (isTypeConstr tab) tyargs)
+                        -- in addition to the arguments explicitly marked for
+                        -- specialisation, also specialise all type arguments
+                        specargs =
+                          nub $
+                            [1 .. tyargsNum]
+                              ++ specargs0
+                    if
+                        | null specargs0 ->
+                            return $ End (mkApps' h args')
+                        | otherwise -> do
+                            eassert (tyargsNum < argsNum)
+                            let def = lookupIdentifierNode tab _identSymbol
+                                (lams, body) = unfoldLambdas def
+                            eassert (length lams == argsNum)
+                            eassert (argsNum <= length tyargs)
+                            let -- We're adding the letrec binder, so need to shift by 1
+                                sargs = map (shift 1) args'
+                                body' =
+                                  replaceIdent _identSymbol argsNum specargs $
+                                    replaceArgs argsNum specargs sargs body
+                                tyargs' = removeSpecTypeArgs specargs sargs (take argsNum tyargs)
+                                tgt' = replaceArgs argsNum specargs sargs (mkPis' (drop argsNum tyargs) tgt)
+                                ty' = mkPis' tyargs' tgt'
+                                lams' =
+                                  zipWithExact
+                                    (\lam ty -> over lambdaLhsBinder (set binderType ty) lam)
+                                    (removeSpecargs specargs lams)
+                                    tyargs'
+                                args'' = removeSpecargs specargs sargs
+                                letitem =
+                                  mkLetItem
+                                    (ii ^. identifierName)
+                                    -- the type is not in the scope of the binder
+                                    (shift (-1) ty')
+                                    (reLambdas lams' body')
+                                node' =
+                                  mkLetRec
+                                    mempty
+                                    (NonEmpty.singleton letitem)
+                                    (mkApps' (mkVar' 0) args'')
+                            node'' <- lambdaLiftNode' True bl node'
+                            return $ End node''
+                  Nothing ->
+                    return $ Recur node
+                where
+                  ii = lookupIdentifierInfo tab _identSymbol
+                  pspec = ii ^. identifierPragmas . pragmasSpecialiseArgs
+                  argsNum = ii ^. identifierArgsNum
+                  (tyargs, tgt) = unfoldPi' (ii ^. identifierType)
+              _ ->
```

Comment:
I think having these dangling `_ -> ...` is bad style, as it makes code harder to read.
Please, refactor the `go` function to improve how it is structured in order to improve readability, or mark this comment as resolved if you think that is not worth it

Reply:
Yes, that's true because the body of the main case is big and you don't see what these wildcards are connected to. I can just put the main case in a separate function.